File: src/ocs_eap_aka.erl
%%% ocs_eap_aka.erl
%%% vim: ts=3
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% @copyright 2019 - 2021 SigScale Global Inc.
%%% @end
%%% Licensed under the Apache License, Version 2.0 (the "License");
%%% you may not use this file except in compliance with the License.
%%% You may obtain a copy of the License at
%%%
%%%     http://www.apache.org/licenses/LICENSE-2.0
%%%
%%% Unless required by applicable law or agreed to in writing, software
%%% distributed under the License is distributed on an "AS IS" BASIS,
%%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%%% See the License for the specific language governing permissions and
%%% limitations under the License.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% @doc This library module implements functions for EAP authentication
%%% 	using 3GPP Authentication and Key Agreement (AKA) in the
%%% 	{@link //ocs. ocs} application.
%%%
%%% @reference <a href="http://tools.ietf.org/html/rfc4187">
%%% 	RFC4187 - EAP Method for 3GPP AKA (EAP-AKA)</a>
%%%
-module(ocs_eap_aka).
-copyright('Copyright (c) 2019 - 2021 SigScale Global Inc.').

% export public api
-export([prf/1, compressed_imsi/1, encrypt_imsi/3, decrypt_imsi/2,
		encrypt_key/4]).

-on_load(init/0).

-include("ocs_eap_codec.hrl").

%% 3GPP TS 23.003 19.3.2 Root NAI
-define(PERM_AKA,  $0).
-define(PERM_AKAp, $6).
%% 3GPP TS 23.003 19.3.4 Fast Re-auth
-define(FAST_AKA,  $4).
-define(FAST_AKAp, $8).
%% 3GPP TS 23.003 19.3.5 Pseudonym
-define(TEMP_AKA,  $2).
-define(TEMP_AKAp, $7).

-spec prf(MK) -> Result
	when 
		MK :: binary(),
		Result :: binary().
%% @doc Pseudo-Random number Function (PRF).
%%
%% 	Generates a digest indistinguisable from random
%% 	as described in RFC4187 section 7.
%%
prf(MK) when byte_size(MK) =:= 20 ->
	prf(MK, <<>>).
%% @hidden
prf(_XKEY, Acc) when byte_size(Acc) =:= 160 ->
	Acc;
prf(<<XKEYn:160>> = XKEY, Acc) ->
	Mod = 1461501637330902918203684832716283019655932542976, % 2^b
	W = g(XKEY),
	<<Wn:160>> = W,
	XKEY1n = (1 + XKEYn + Wn) rem Mod,
	prf(<<XKEY1n:160>>, <<Acc/binary, W/binary>>).

%%
%% internal functions
%% 

-spec g(XKEY) -> Digest
	when
		XKEY :: binary(),
		Digest:: binary().
%% @doc Calculate message digest.
%%
%% 	A modified SHA-1 as described in RFC4187 section 7.
g(XKEY) when bit_size(XKEY) =:= 160 ->
	erlang:nif_error(nif_library_not_loaded).

-spec compressed_imsi(IMSI) -> IMSI
	when
		IMSI :: binary().
%% @doc Compress or decompress an IMSI.
%%
%% 	See 3GPP 33.402 14.1 Temporary identity generation.
%% @private
compressed_imsi(<<15:4, _:60/bits>> = IMSI) ->
	B = << <<A, B>> || <<A:4, B:4>> <= IMSI >>,
	lists:flatten([integer_to_list(C) || <<C>> <= B, C /= 15]);
compressed_imsi(IMSI) when is_binary(IMSI) ->
	L1 = [list_to_integer([C]) || <<C>> <= IMSI],
	L2 = lists:duplicate(16 - length(L1), 15),
	L3 = L2 ++ L1,
	<< <<D:4>> || D <- L3 >>.

-dialyzer([{nowarn_function, [encrypt_imsi/3]}, no_missing_calls]). % temporary
-spec encrypt_imsi(Tag, CompressedIMSI, Key) -> Pseudonym
	when
		Tag :: 50 | 55, % ?TEMP_AKA | ?TEMP_AKAp, (compile error with 18)
		CompressedIMSI :: binary(),
		Key :: {N, Kpseu},
		N :: pos_integer(),
		Kpseu :: binary(),
		Pseudonym :: binary().
%% @doc Create an encrypted temporary identity.
%%
%% 	See 3GPP 33.402 14.1 Temporary identity generation.
%% @private
encrypt_imsi(Tag, CompressedIMSI, {N, Kpseu} = _Key)
		when ((Tag =:= ?TEMP_AKA) or (Tag =:= ?TEMP_AKAp)),
		size(CompressedIMSI) == 8, size(Kpseu) == 16 ->
	Pad = crypto:strong_rand_bytes(8),
	PaddedIMSI = <<CompressedIMSI/binary, Pad/binary>>,
	EncryptedIMSI = crypto:block_encrypt(aes_ecb, Kpseu, PaddedIMSI),
	TaggedIMSI = <<Tag:6, N:4, EncryptedIMSI/binary, 0:6>>,
	binary:part(base64:encode(TaggedIMSI), 0, 23).

-spec decrypt_imsi(Pseudonym, Keys) -> CompressedIMSI
	when
		Pseudonym :: binary(),
		Keys :: [Key],
		Key :: {N, Kpseu},
		N :: pos_integer(),
		Kpseu :: binary(),
		CompressedIMSI :: binary().
%% @doc Decrypt a temporary identity.
%%
%% 	See 3GPP 33.402 14.1 Temporary identity generation.
%% @private
decrypt_imsi(Pseudonym, Keys)
		when size(Pseudonym) == 23, is_list(Keys) ->
	TaggedIMSI = base64:decode(<<Pseudonym/binary, $A>>),
	<<_:6, N:4, EncryptedIMSI:16/binary, _:6>> = TaggedIMSI,
	{_, Kpseu} = lists:keyfind(N, 1, Keys),
	PaddedIMSI = crypto:block_decrypt(aes_ecb, Kpseu, EncryptedIMSI),
	binary:part(PaddedIMSI, 0, 8).

-spec encrypt_key(Secret, RequestAuthenticator, Salt, Key) -> Ciphertext
	when
		Secret :: binary(),
		RequestAuthenticator :: [byte()],
		Salt :: integer(),
		Key :: binary(),
		Ciphertext :: binary().
%% @doc Encrypt the Pairwise Master Key (PMK) according to RFC2548
%% 	section 2.4.2 for use as String in a MS-MPPE-Recv-Key
%% 	or MS-MPPE-Send-Key attribute.
%% @private
encrypt_key(Secret, RequestAuthenticator, Salt, Key)
		when (Salt bsr 15) == 1 ->
	KeyLength = size(Key),
	Plaintext = case (KeyLength + 1) rem 16 of
		0 ->
			<<KeyLength, Key/binary>>;
		N ->
			PadLength = (16 - N) * 8,
			<<KeyLength, Key/binary, 0:PadLength>>
	end,
	F = fun(P, [H | _] = Acc) ->
				B = crypto:hash(md5, [Secret, H]),
				C = crypto:exor(P, B),
				[C | Acc]
	end,
	AccIn = [[RequestAuthenticator, <<Salt:16>>]],
	AccOut = lists:foldl(F, AccIn, [P || <<P:16/binary>> <= Plaintext]),
	iolist_to_binary(tl(lists:reverse(AccOut))).

%%
%% internal functions
%%

-spec init() -> ok.
%% @doc When this module is loaded this function is called to load NIF library.
%% @hidden
init() ->
	{ok, Application} = application:get_application(?MODULE),
	PrivDir = case code:priv_dir(Application) of
		{error, bad_name} ->
			BEAM = atom_to_list(?MODULE) ++ ".beam",
			Ebin = filename:dirname(code:where_is_file(BEAM)),
			filename:dirname(Ebin) ++ "/priv";
		Path ->
			Path
	end,
	ok = erlang:load_nif(PrivDir ++ "/lib/ocs_eap_aka", 0).

File: c_src/ocs_eap_aka.c
/* ocs_eap_aka.c
*** vim: ts=3
*****************************************************************************
*** Copyright 2016 - 2017 SigScale Global Inc.
*** 
*** Licensed under the Apache License, Version 2.0 (the "License");
*** you may not use this file except in compliance with the License.
*** You may obtain a copy of the License at
***
***     http://www.apache.org/licenses/LICENSE-2.0
***
*** Unless required by applicable law or agreed to in writing, software
*** distributed under the License is distributed on an "AS IS" BASIS,
*** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*** See the License for the specific language governing permissions and
*** limitations under the License.
*****************************************************************************
*** This module implements NIFs for the modified SHA-1 used by
*** EAP-AKA (RFC4187).
***/

#include <stdint.h>
#include "erl_nif.h"

#if ((ERL_NIF_MAJOR_VERSION == 2 && ERL_NIF_MINOR_VERSION < 8) \
		|| ERL_NIF_MAJOR_VERSION < 2)
ERL_NIF_TERM
enif_raise_exception(ErlNifEnv* env, ERL_NIF_TERM reason) {
	return enif_make_badarg(env);
}
#endif /* NIF < v2.8 */

#define CLS(B, W) ((W << B) | (W >> (32 - B)))
#define F1(B, C, D) ((B & C) | ((~B) & D))
#define F2(B, C, D) (B ^ C ^ D)
#define F3(B, C, D) ((B & C) | (B & D) | (C & D))

static ERL_NIF_TERM
g_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
   ErlNifBinary xkey, res;
   ERL_NIF_TERM reason;
   int i;
   const uint32_t h[5] = {0x67452301,
			0xefcdab89, 0x98badcfe, 0x10325476, 0xc3d2e1f0};
   uint32_t a, b, c, d, e, temp;
   uint32_t m[16] = {0};
   uint32_t w[80] = {0};
   
   if (!enif_inspect_binary(env, argv[0], &xkey) || xkey.size != 20)
      return enif_make_badarg(env);
   for(i = 0; i < 5; i++) {
      m[i] = (xkey.data[i*4] << 24) | (xkey.data[i*4+1] << 16)
            | (xkey.data[i*4+2] << 8) | (xkey.data[i*4+3]);
   } 
	for(i = 0; i < 16; i++) {
		w[i] = m[i];
	}
	for(i = 16; i < 80; i++) {
		w[i] = CLS(1, w[i-3] ^ w[i-8] ^ w[i-14] ^ w[i-16]);
	}
	a = h[0];
	b = h[1];
	c = h[2];
	d = h[3];
	e = h[4];
	for(i = 0; i < 20; i++) {
		temp = CLS(5, a) + F1(a, b, c) + e + w[i] + 0x5aA827999;
		e = d;
		d = c;
		c = CLS(30, b);
		b = a;
		a = temp;

	}
	for(i = 20; i < 40; i++) {
		temp = CLS(5, a) + F2(a, b, c) + e + w[i] + 0x6ed9eba1;
		e = d;
		d = c;
		c = CLS(30, b);
		b = a;
		a = temp;
	}
	for(i = 40; i < 60; i++) {
		temp = CLS(5, a) + F3(a, b, c) + e + w[i] + 0x8f1bbcdc;
		e = d;
		d = c;
		c = CLS(30, b);
		b = a;
		a = temp;
	}
	for(i = 60; i < 80; i++) {
		temp = CLS(5, a) + F2(a, b, c) + e + w[i] + 0xca62c1d6;
		e = d;
		d = c;
		c = CLS(30, b);
		b = a;
		a = temp;
	}
	if (!enif_alloc_binary(20, &res)) {
		enif_make_existing_atom(env, "enomem", &reason, ERL_NIF_LATIN1);
		return enif_raise_exception(env, reason);
	}
	temp = h[0] + a;
	res.data[0] = temp >> 24;
	res.data[1] = (temp >> 16) & 0xff;
	res.data[2] = (temp >> 8) & 0xff;
	res.data[3] = temp & 0xff;
	temp = h[1] + b;
	res.data[4] = temp >> 24;
	res.data[5] = (temp >> 16) & 0xff;
	res.data[6] = (temp >> 8) & 0xff;
	res.data[7] = temp & 0xff;
	temp = h[2] + c;
	res.data[8] = temp >> 24;
	res.data[9] = (temp >> 16) & 0xff;
	res.data[10] = (temp >> 8) & 0xff;
	res.data[11] = temp & 0xff;
	temp = h[3] + d;
	res.data[12] = temp >> 24;
	res.data[13] = (temp >> 16) & 0xff;
	res.data[14] = (temp >> 8) & 0xff;
	res.data[15] = temp & 0xff;
	temp = h[4] + e;
	res.data[16] = temp >> 24;
	res.data[17] = (temp >> 16) & 0xff;
	res.data[18] = (temp >> 8) & 0xff;
	res.data[19] = temp & 0xff;
	return enif_make_binary(env, &res);
}

static ErlNifFunc
nif_funcs[] = {
   {"g", 1, g_nif}
};

ERL_NIF_INIT(ocs_eap_aka, nif_funcs, NULL, NULL, NULL, NULL)

